File: deque/core_problems.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

// Problem 1: Sliding Window Maximum
vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    deque<int> dq;
    vector<int> result;
    
    for (int i = 0; i < nums.size(); i++) {
        while (!dq.empty() && dq.front() < i - k + 1) {
            dq.pop_front();
        }
        
        while (!dq.empty() && nums[dq.back()] < nums[i]) {
            dq.pop_back();
        }
        
        dq.push_back(i);
        
        if (i >= k - 1) {
            result.push_back(nums[dq.front()]);
        }
    }
    
    return result;
}

// Problem 2: Palindrome Check
bool isPalindrome(string s) {
    deque<char> dq;
    
    for (char c : s) {
        if (isalnum(c)) {
            dq.push_back(tolower(c));
        }
    }
    
    while (dq.size() > 1) {
        if (dq.front() != dq.back()) {
            return false;
        }
        dq.pop_front();
        dq.pop_back();
    }
    
    return true;
}

int main() {
    cout << "=== Core Deque Problems ===" << endl << endl;
    
    // Test 1
    cout << "Test 1: Sliding Window Maximum" << endl;
    vector<int> nums = {1, 3, -1, -3, 5, 3, 6, 7};
    int k = 3;
    vector<int> result = maxSlidingWindow(nums, k);
    cout << "Result: [ ";
    for (int r : result) cout << r << " ";
    cout << "]" << endl << endl;
    
    // Test 2
    cout << "Test 2: Palindrome Check" << endl;
    cout << "'racecar' is palindrome: " << (isPalindrome("racecar") ? "true" : "false") << endl;
    cout << "'hello' is palindrome: " << (isPalindrome("hello") ? "true" : "false") << endl;
    
    return 0;
}

File: Level 1 - Foundations/deque/core_problems.java
package deque;

import java.util.*;

public class core_problems {

    // Problem 1: Sliding Window Maximum using Deque
    static int[] maxSlidingWindow(int[] nums, int k) {
        if (nums == null || k == 0)
            return new int[0];

        Deque<Integer> deque = new LinkedList<>();
        int[] result = new int[nums.length - k + 1];

        for (int i = 0; i < nums.length; i++) {
            // Remove elements outside window
            while (!deque.isEmpty() && deque.peekFirst() < i - k + 1) {
                deque.pollFirst();
            }

            // Remove smaller elements (they won't be max)
            while (!deque.isEmpty() && nums[deque.peekLast()] < nums[i]) {
                deque.pollLast();
            }

            deque.offerLast(i);

            if (i >= k - 1) {
                result[i - k + 1] = nums[deque.peekFirst()];
            }
        }

        return result;
    }

    // Problem 2: Implement Queue using Deque
    static class QueueUsingDeque {
        Deque<Integer> deque = new LinkedList<>();

        void enqueue(int val) {
            deque.offerLast(val);
        }

        int dequeue() {
            return deque.isEmpty() ? -1 : deque.pollFirst();
        }

        int front() {
            return deque.isEmpty() ? -1 : deque.peekFirst();
        }
    }

    // Problem 3: Implement Stack using Deque
    static class StackUsingDeque {
        Deque<Integer> deque = new LinkedList<>();

        void push(int val) {
            deque.offerLast(val);
        }

        int pop() {
            return deque.isEmpty() ? -1 : deque.pollLast();
        }

        int top() {
            return deque.isEmpty() ? -1 : deque.peekLast();
        }
    }

    // Problem 4: Check if Palindrome using Deque
    static boolean isPalindrome(String s) {
        Deque<Character> deque = new LinkedList<>();

        for (char c : s.toLowerCase().toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                deque.offerLast(c);
            }
        }

        while (deque.size() > 1) {
            if (deque.pollFirst() != deque.pollLast()) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        System.out.println("=== Core Deque Problems ===\n");

        // Test 1: Sliding Window Maximum
        System.out.println("Test 1: Sliding Window Maximum");
        int[] nums = { 1, 3, -1, -3, 5, 3, 6, 7 };
        int k = 3;
        int[] result = maxSlidingWindow(nums, k);
        System.out.println("Array: " + Arrays.toString(nums));
        System.out.println("k = " + k);
        System.out.println("Result: " + Arrays.toString(result));
        System.out.println();

        // Test 2: Queue using Deque
        System.out.println("Test 2: Queue using Deque");
        QueueUsingDeque queue = new QueueUsingDeque();
        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);
        System.out.println("Dequeue: " + queue.dequeue());
        System.out.println("Front: " + queue.front());
        System.out.println();

        // Test 3: Palindrome Check
        System.out.println("Test 3: Palindrome Check");
        System.out.println("'racecar' is palindrome: " + isPalindrome("racecar"));
        System.out.println("'hello' is palindrome: " + isPalindrome("hello"));
    }
}
